common/dynamicconfig/collection.go

```diff
@@ -0,0 +1,543 @@
+	}
+
+	// These function types follow a similar pattern:
+	//   {X}PropertyFn - returns a value of type X that is global (no filters)
+	//   {X}PropertyFnWith{Y}Filter - returns a value of type X with the given filters
+	// Available value types:
+	//   Bool: bool
+	//   Duration: time.Duration
+	//   Float: float64
+	//   Int: int
+	//   Map: map[string]any
+	//   String: string
+	// Available filters:
+	//   Namespace func(namespace string)
+	//   NamespaceID func(namespaceID string)
+	//   TaskQueueInfo func(namespace string, taskQueue string, taskType enumspb.TaskQueueType)
+	//   ShardID func(shardID int32)
+	BoolPropertyFn                             func() bool
+	BoolPropertyFnWithNamespaceFilter          func(namespace string) bool
+	BoolPropertyFnWithNamespaceIDFilter        func(namespaceID string) bool
+	BoolPropertyFnWithTaskQueueInfoFilters     func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) bool
+	DurationPropertyFn                         func() time.Duration
+	DurationPropertyFnWithNamespaceFilter      func(namespace string) time.Duration
+	DurationPropertyFnWithNamespaceIDFilter    func(namespaceID string) time.Duration
+	DurationPropertyFnWithShardIDFilter        func(shardID int32) time.Duration
+	DurationPropertyFnWithTaskQueueInfoFilters func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) time.Duration
+	FloatPropertyFn                            func() float64
+	FloatPropertyFnWithNamespaceFilter         func(namespace string) float64
+	FloatPropertyFnWithShardIDFilter           func(shardID int32) float64
+	FloatPropertyFnWithTaskQueueInfoFilters    func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) float64
+	IntPropertyFn                              func() int
+	IntPropertyFnWithNamespaceFilter           func(namespace string) int
+	IntPropertyFnWithShardIDFilter             func(shardID int32) int
+	IntPropertyFnWithTaskQueueInfoFilters      func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) int
+	MapPropertyFn                              func() map[string]any
+	MapPropertyFnWithNamespaceFilter           func(namespace string) map[string]any
+	StringPropertyFn                           func() string
+	StringPropertyFnWithNamespaceFilter        func(namespace string) string
+)
+
+const (
+	errCountLogThreshold = 1000
+)
+
+var (
+	errKeyNotPresent        = errors.New("key not present")
+	errNoMatchingConstraint = errors.New("no matching constraint in key")
+)
+
+// NewCollection creates a new collection
+func NewCollection(source Source, logger log.Logger) *Collection {
+	return &Collection{
+		source:   source,
+		logger:   logger,
+		errCount: -1,
+	}
+}
+
+func (c *Collection) logError(key Key, err error) {
+	// TODO: This is a lot of unnecessary contention with little benefit. Consider using
+	// https://github.com/cespare/percpu here.
+	errCount := atomic.AddInt64(&c.errCount, 1)
+	if errCount%errCountLogThreshold == 0 {
```

Comment:
Ideally we want to sample log per key, and we want to log for first encounter. But for now, maybe change errCount%errCountLogThreshold == 1 so we at least log first one.

Reply:
see above, errCount starts at -1.

for doing a sample per key, I think a sync.Map could work? I'm not convinced any of this is worth it though. most dynamic config values will be unset so it's a lot of noise